[WIP] Build sap portal

polybuildr opened the pull request:
Work in progress, do not merge! But please review and leave comments. Meant to resolve T83.

polybuildr committed:
Change fields from mediumtext to text in sap_ambassadors table

polybuildr committed:
Update ThodaCSS, also include thoda.min.js

polybuildr committed:
Move SAP navbar to a fragment

polybuildr committed:
Add users and user login to SAP portal

polybuildr commented:
Partially {resolves T83}.

polybuildr committed:
Add missions and tasks to SAP, /sap/mission/<id> to view a mission

polybuildr committed:
Make sap/navbar_fragment intelligent wrt logged_in status

If logged in, show different links, else show log in link.

polybuildr committed:
Add intelligent navbar to SAP views

polybuildr committed:
Display missions on SAP dashboard

polybuildr committed:
Allow admins to create new missions

Creation of tasks is not supported yet.

nisargjhaveri commented on src/app/views/sap/register.php:
```diff
@@ -9,14 +9,7 @@
-        <a href="javascript:void(0)" class="menu-toggle-button">&equiv;</a>
-        </a>
-    </div>
+    <?php $this->load_fragment('sap/navbar_fragment', ['logged_in' => $logged_in]); ?>
```
It already passes whatever is passed to the view. So, I don't think you need to pass `logged_in` again.

nisargjhaveri replied:
Well, it is more easier to comprehend this way. So, don't know if we should keep it or not.

polybuildr replied:
I didn't know that. I was wishing that the functionality was there, so I could use it. :P But you're right, it is easier to understand this way. Let's keep it like this until something changes.

polybuildr committed:
Update thoda.min.css, adds margin to <li> elements

polybuildr committed:
Add create task functionality for admins

- /sap/portal/mission/<id>/createtask lets admins create tasks
- Edited DB schema for tasks, `text_answer` -> `has_text_answer`
- Refactored sap_portal controller, added and used a validate_data
  method

polybuildr committed:
Allow users to submit tasks for review

polybuildr committed:
Refactor sap_portal task submission to use flash data for errors

polybuildr committed:
SAP: Change get_task_submissions to get from a particular mission

polybuildr committed:
Move submission and task matching to sap_model

polybuildr committed:
Fix two minor spacing issues

polybuildr committed:
Add sap_task_submissions database schema

polybuildr committed:
Allow admins to review tasks, show review status to users

polybuildr commented:
Okay, I think I'm going to have to temporarily stop work on this now and hand it over to you guys, @nisargjhaveri, @hharchani. We need to get this deployed in the next 20 hours, during which time I won't be available for most of it.

A lot of things are not done, but these are the basics, and it should not be too difficult to build on top of this. Please merge this and implement on top of it, or of course, pull from my branch and edit and commit and new PR.

One major issue is that users don't get added to `sap_users` in any way, so you'll have to manually create users to test anything. (You can do the initial registration using the form, though.)

Things that work:
- Admins can create missions and create tasks within those missions. These tasks can also optionally include filling of a text field.
- Users can submit tasks, including filling the text field.
- Admins can review tasks,  "approve" or "reject"
- Users' view of a mission shows accepted tasks, also shows rejected tasks only first page load (sorry)
- Users can resubmit rejected tasks

Things that must be done:
- Awarding of scores. The `sap_portal` controller has a `TODO`. Once all tasks of a mission are completed, the user's score must increase by the amount specified in the mission. The score should also be displayed somewhere.
- A view where admins can view users who have registered, and "approve" them to join the sap (technically, copy users from `sap_ambassadors` to `sap_users`. Please look into removing the "username" field and replacing it with the email and additionally, the name or something. The username is currently only being used for logins.
- Database schema checking. At least one foreign key is missing. `registration_id` in `sap_user`, I think. Some `UNIQUE` checks are missing. While the UI does not support it, it is possible for an ambassador to make multiple task submissions at the moment.

Things that should be done:
- Leaderboard. Ranking of people by scores.
- Duplication of some fields across tables, with `ON UPDATE CASCADE`. Refer line 181 of `sap_model` in the current version of this PR.
- Lots more. Some maybe should go into the above category. :P I'm sorry, it's late, I'm tired and I'm also tired of working on this portal. Help!

polybuildr commented:
More 'must's.
- `htmlspecialchars` some stuff that I'm printing without sanitizing

One thing to keep in mind is that you can trust SAP admins to not game the system, but not ambassadors - since we don't have any character checking of the ambassadors.

polybuildr committed:
Added some htmlspecialchars() to SAP views to sanitize user input

polybuildr commented:
Attempted to fix the unsanitized output since we **really** don't want that.

polybuildr commented:
Additional features that _must_ be there:
- Deleting a mission or a task

Additional features that _should_ be there:
- Editing/updating a mission or a task

hharchani commented on src/app/controllers/sap.php:
```diff
@@ -34,10 +36,40 @@ function index() {
+
+    public function logout() {
+        $this->load_library('sap_auth_lib', 'sap_auth');
+        $this->load_library('http_lib');
```
Why is this loaded?

polybuildr replied:
My bad, no reason.

polybuildr committed:
Remove useless line from sap::logout

polybuildr committed:
SAP: Add comments and do minor refactoring to clean code

polybuildr committed:
Default mission id to null in sap_portal:mission

This way /sap/mission won't show an E_WARNING because $id is unset.

hharchani commented on src/app/models/sap_model.php:
```diff
@@ -22,4 +24,220 @@ public function registerEntry($data) {
+ON `ambassadors`.`id` = `users`.`registration_id`
+WHERE `tasks`.`mission_id`=? and `task_submissions`.`done` = 0
+SQL;
+        $stmt = $this->db_lib->prepared_execute(
```
Itni gandi query kyu hai? :( :disappointed:

polybuildr replied:
Look at line 181. :P

polybuildr commented:
Rebased on master after #74 and also fixed a typo in the last commit message.

polybuildr committed:
Change username to email all over SAP

The username field is useless, and might be unintuitive in certain
cases. Remove all instances of it and replace it with an email ID.

polybuildr committed:
Remove intermediate migrations and keep only required ones, with prefix

The kept migrations have an XXXX numeric prefix.

hharchani commented:
Looks fine. Merging. @smritz will test it afterwards.
